toolchain/semantics/semantics_handle_index.cpp

```diff
@@ -15,79 +15,89 @@ auto HandleIndexExpressionStart(Context& /*context*/,
   return true;
 }
 
+// Validates that the index (required to be an IntegerLiteral) is valid within
+// the array or tuple size. Returns the index on success, or nullptr on failure.
+static auto ValidateIntegerLiteralBound(Context& context,
+                                        ParseTree::Node parse_node,
+                                        SemIR::Node operand_node,
+                                        SemIR::Node index_node, int size)
+    -> const llvm::APInt* {
+  const auto& index_val = context.semantics_ir().GetIntegerLiteral(
+      index_node.GetAsIntegerLiteral());
+  if (index_val.uge(size)) {
+    CARBON_DIAGNOSTIC(IndexOutOfBounds, Error,
+                      "Index `{0}` is past the end of `{1}`.", llvm::APSInt,
+                      std::string);
+    context.emitter().Emit(
+        parse_node, IndexOutOfBounds,
+        llvm::APSInt(index_val, /*isUnsigned=*/true),
+        context.semantics_ir().StringifyType(operand_node.type_id()));
+    return nullptr;
+  }
+  return &index_val;
+}
+
 auto HandleIndexExpression(Context& context, ParseTree::Node parse_node)
     -> bool {
-  CARBON_DIAGNOSTIC(OutOfBoundsAccess, Error,
-                    "Index `{0}` is past the end of `{1}`.", llvm::APSInt,
-                    std::string);
-
   auto index_node_id = context.node_stack().PopExpression();
   auto index_node = context.semantics_ir().GetNode(index_node_id);
-  auto name_node_id = context.node_stack().PopExpression();
-  name_node_id = context.MaterializeIfInitializing(name_node_id);
-  auto name_node = context.semantics_ir().GetNode(name_node_id);
-  auto name_type_id =
-      context.semantics_ir().GetTypeAllowBuiltinTypes(name_node.type_id());
-  auto name_type_node = context.semantics_ir().GetNode(name_type_id);
+  auto operand_node_id = context.node_stack().PopExpression();
+  operand_node_id = context.MaterializeIfInitializing(operand_node_id);
```

Comment:
I'm OK with `operand`, but another candidate would be `indexed`, looking like:

```suggestion
  auto indexed_node_id = context.node_stack().PopExpression();
  indexed_node_id = context.MaterializeIfInitializing(indexed_node_id);
```

While I might prefer `indexed`/`index` vs `operand`/`index` by a slight margin, its really slight, so go with whichever works best for you.

Reply:
I'd started with `indexed`, but it's visually very close to `index` and was hard for me to read. `operand` was my choice to make it easy to see one versus the other.

Reply:
Yeah. Long-term, I think we want somehow to model both of these: that it *is* an error, and the IR that would have been there anyways. Right now, putting in the error breaks the connection in the IR.

The reason I think we'll want this is for tools that are trying to reason about the SemIR in the face of errors (for example to suggest fixes, etc) -- I think they'll want a bit more information.

But either way (with the error none in your latest push, or the original form) is fine for now, this isn't blocking at all.

Reply:
SGTM, although I think you meant to put this comment on the other thread. ;)

Reply:
Doh, yes, sorry.